### PR TITLE
Fix #17600 - Tuplets Skipped by Alt+left and Alt+right

### LINKED ISSUE #17600: [Accessibility] Tuplets skipped by Alt+Left and Alt+Right
### Issue type

UI bug

### Bug description

The `next-element` (Alt+Right) and `previous-element` (Alt+Left) actions are able to select notes within a tuplet, however the tuplet itself cannot be selected.

### Steps to reproduce

1. Create a score
2. Select a rest and press Ctrl+3 to add a tuplet.
3. Select a nearby note or rest.
4. Try to navigate to the tuplet with Alt+Left or Alt+Right.

### Screenshots/Screen recordings

image

### MuseScore Version

Latest nightly (231340304, 0e5a099)

### Regression

No.

### Operating system

Windows 11

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by codecruisedor**
Resolves: #17600

I have changed four files in this PR. This is my first attempt at raising a pull request, so please review the code and suggest changes if needed. Thanks.!!

- [X] I signed the [CLA](https://musescore.org/en/cla)
- [X] The title of the PR describes the problem it addresses
- [X] Each commit's message describes its purpose and effects, and references the issue it resolves
- [X] If changes are extensive, there is a sequence of easily reviewable commits
- [X] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [X] There are no unnecessary changes
- [X] The code compiles and runs on my machine, preferably after each commit individually

### CODE REVIEW

File: src/engraving/libmscore/chordrest.cpp [hunk 1]
```diff
@@ -917,6 +917,9 @@ EngravingItem* ChordRest::prevArticulationOrLyric(EngravingItem* e)
 
 EngravingItem* ChordRest::nextElement()
 {
+    if (nextChordRest(this)->tuplet()) {
+        return nextChordRest(this)->tuplet();
```
Comment (mike-spa) [inline src/engraving/libmscore/chordrest.cpp:921]: Thank you! A couple of things:
1. Please don't call `nextChordRest` twice, it's quite a big function. Just store the result in a temp variable the first time.
2. `nextChordRest` may return null, so please null-check the output of the function before using it.

File: src/engraving/libmscore/chordrest.cpp [hunk 2]
```diff
@@ -951,6 +954,9 @@ EngravingItem* ChordRest::nextElement()
 
 EngravingItem* ChordRest::prevElement()
 {
+    if (prevChordRest(this)->tuplet()) {
+        return prevChordRest(this)->tuplet();
+    }
```
Comment (mike-spa) [inline src/engraving/libmscore/chordrest.cpp:959]: Same as above

**Comment by codecruisedor** [comment]
Hello @mike-spa, thanks for the review. I have made the changes as you prescribed in the latest commit. The Generate and compare test for this PR is failing after my latest commit, I don't understand why. Can you please shed some light on it?

**Comment by Jojo-Schmitz** [comment]
I don't see how the failiong vtests rmay relate to the changes in this PR:
articulation-beams-1-1 diff
articulation-beams-2-1 diff
rests-6-1 diff
voice-2-1 diff
But they are quite bad, esp. the last 2

**Comment by cbjeukendrup** [comment]
A rebase should fix it. The changes were caused by different PRs, but if your PR's branch is not based upon the latest master branch, the VTests still get upset.

File: src/engraving/libmscore/chordrest.cpp [hunk 3]
```diff
@@ -917,6 +917,10 @@ EngravingItem* ChordRest::prevArticulationOrLyric(EngravingItem* e)
 
 EngravingItem* ChordRest::nextElement()
 {
+    Tuplet* tuplet = nextChordRest(this)->tuplet();
```
Comment (mike-spa) [inline src/engraving/libmscore/chordrest.cpp:920]: Almost there! You still need to null-check the result of `nextChordRest(this)`, before you try to `->tuplet()` on it. You can write like this:
```
ChordRest* nextCR = nextChordRest(this);
Tuplet* = nextCR ? nextCR->tuplet() : nullptr;
if (tuplet) {
    return tuplet;
}
```

Same goes for `prevChordRest(this)` below.

**Comment by mike-spa** [comment]
> Hello @mike-spa, thanks for the review. I have made the changes as you prescribed in the latest commit. The Generate and compare test for this PR is failing after my latest commit, I don't understand why. Can you please shed some light on it?

Yep, like @cbjeukendrup said, those are other fixes that went into master while your PR was open. Rebasing you branch on latest master will fix it :)

**Review by mike-spa — APPROVED** [review]

**Comment by codecruisedor** [comment]
Hello guys, has anyone tested these changes, are they working as intended? Also, do I need to squash my commits before they can be merged?

**Comment by cbjeukendrup** [comment]
This will be tested soon! The testers are currently busy with the 4.1 release.

Squashing the commits would be good indeed, in this case. It would be great if you could also rebase the branch on top of the latest master branch (really rebase instead of merge). Thanks!

**Comment by codecruisedor** [comment]
Okay, so should I squash and rebase now or wait till the testing is complete for these changes?

**Comment by cbjeukendrup** [comment]
@codecruisedor Yes, that would be great!

**Comment by cbjeukendrup** [comment]
@codecruisedor It looks like squashing and rebasing didn't really go well... we can squash the commits for you when merging, but I would like to offer you the opportunity to learn how to do it, since being comfortable with Git is crucial in life :) 

The information below assumes that your `upstream` remote points to the musescore/MuseScore repository. 

The easiest way to simply rebase your branch, is:
```
git pull --rebase upstream master
```
Note the --rebase flag. Also note that _rebasing_ and _merging_ are different things. For PR branches, rebasing is preferred. 

But you can also squash and rebase in one action. First, download the latest changes from upstream:
```
git fetch upstream master
```
Then, rebase your branch "interactively":
```
git rebase -i upstream/master
```
Then, a text editor will show up. In this editor, you see a list of all your commits. Leave the line of the first commit (topmost) untouched. On the subsequent lines, replace "pick" with "squash", as explained by Git itself. 

After doing this, save and close the text file. Git will do it's work, and when it's finished, you can force-push your changes to GitHub:
```
git push -f
```

One more thing: if you have never customised your Git text editor, you will most likely get Vim. My personal experience is that it's unlikely that you figure out how that works by just using your intuition, but there are plenty of Vim tutorials online. And also plenty of tutorials on how to change the Git text editor. 

I hope this info helps!

**Comment by codecruisedor** [comment]
Thanks a lot for this info, @cbjeukendrup. You sensed correctly that I was struggling with rebase. For some reason doing interactive rebase via git bash did not seem to work. I will invest some time in properly learning git. Thanks again for your help.:)

**Comment by cbjeukendrup** [comment]
Hi @codecruisedor! Could you please give us an update about the progress with this one?

**Comment by codecruisedor** [comment]
Hi @cbjeukendrup , I tried to rebase and squash today..it did squash it to 1 commit. But somehow it incorporated other changes into this pr I think. Would it be better if I opened another pr with the same changes?

**Comment by cbjeukendrup** [comment]
Yes, perhaps that's the best idea at this point. Thanks!

**Comment by cbjeukendrup** [comment]
(Closing as superseded by https://github.com/musescore/MuseScore/pull/18947)

**Pull request closed without merging**